From 79c03392966237689f478d41461726e67a0dc228 Mon Sep 17 00:00:00 2001
From: Takanari Hayama <taki@igel.co.jp>
Date: Fri, 20 Mar 2015 11:04:00 +0900
Subject: [PATCH] v4l2-renderer: release DMABUF handle when closing KMS BOs.

Just release KMS BO doesn't actually release underlying buffer. We shall
close DMABUF handle in order to release the buffer. This commit fixes
the memory leak.

Reported-By: Kazunori Kobayashi <kkobayas@igel.co.jp>
---
 src/v4l2-renderer.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/src/v4l2-renderer.c b/src/v4l2-renderer.c
index d15bf75..f7b0d97 100644
--- a/src/v4l2-renderer.c
+++ b/src/v4l2-renderer.c
@@ -36,6 +36,7 @@
 #include <sys/types.h>
 #include <sys/stat.h>
 #include <fcntl.h>
+#include <unistd.h>
 
 #include <linux/videodev2.h>
 #include <linux/v4l2-subdev.h>
@@ -885,10 +886,19 @@ v4l2_renderer_flush_damage(struct weston_surface *surface)
 static void
 v4l2_release_kms_bo(struct v4l2_surface_state *vs)
 {
+	int i;
+
 	if (!vs)
 		return;
 
 	if (vs->bo) {
+		for (i = 0; i < vs->num_planes; i++) {
+			if (vs->planes[i].dmafd >= 0) {
+				close(vs->planes[i].dmafd);
+				vs->planes[i].dmafd = -1;
+			}
+		}
+
 		if (kms_bo_unmap(vs->bo))
 			weston_log("kms_bo_unmap failed.\n");
 
@@ -972,6 +982,7 @@ v4l2_renderer_attach_shm(struct v4l2_surface_state *vs, struct weston_buffer *bu
 	vs->pixel_format = pixel_format;
 	vs->num_planes = 1;
 	vs->planes[0].stride = stride;
+	vs->planes[0].dmafd = -1;
 	vs->bpp = bpp;
 
 	if (device_interface->attach_buffer(vs) == -1)
-- 
1.9.1

